docs/proposals/rescheduling-for-critical-pods.md

```diff
@@ -0,0 +1,116 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Rescheduler: guaranteed scheduling of critical addons
+
+## Motivation
+
+In addition to Kubernetes core components like api-server, scheduler, controller-manager running on a master machine
+there is a bunch of addons which due to various reasons have to run on a regular cluster node, not the master.
+Some of them are critical to have fully functional cluster: Heapster, DNS, UI. Users can break their cluster
+by evicting a critical addon (either manually or as a side effect of an other operation like upgrade)
+which possibly can become pending (for example when the cluster is highly utilized).
+To avoid such situation we want to have a mechanism which guarantees that
+critical addons are scheduled assuming the cluster is big enough.
+This possibly may affect other pods (including production user’s applications).
+
+## Design
+
+Rescheduler will ensure that critical addons are always scheduled.
+In the first version it will implement only this policy, but later we may want to introduce other policies.
+It will be a standalone component running on master machine similarly to scheduler.
+Those components will share common logic (initially rescheduler will in fact import some of scheduler packages).
+
+### Guaranteed scheduling of critical addons
+
+Rescheduler will observe critical addons (running in `kube-system` namespace
+with annotation `alpha.kubernetes.io/service-type` set to `CRITICAL`).
+If one of them is marked by scheduler as unschedulable (pod condition `PodScheduled` set to `false`, the reason set to `Unschedulable`)
+the component will try to find a space for the addon by evicting some pods and then the scheduler will schedule the addon.
```

Comment:
ACk I didn't read ahead

On Jul 24, 2016 12:18 PM, "David Oppenheimer" <notifications@github.com>
wrote:

> In docs/proposals/rescheduling-for-critical-pods.md
> <https://github.com/kubernetes/kubernetes/pull/29195#discussion_r71993725>
> :
>
> > +critical addons are scheduled assuming the cluster is big enough.
> > +This possibly may affect other pods (including production user’s applications).
> > +
> > +## Design
> > +
> > +Rescheduler will ensure that critical addons are always scheduled.
> > +In the first version it will implement only this policy, but later we may want to introduce other policies.
> > +It will be a standalone component running on master machine similarly to scheduler.
> > +Those components will share common logic (initially rescheduler will in fact import some of scheduler packages).
> > +
> > +### Guaranteed scheduling of critical addons
> > +
> > +Rescheduler will observe critical addons (running in `kube-system` namespace
> > +with annotation `alpha.kubernetes.io/service-type` <http://alpha.kubernetes.io/service-type> set to `CRITICAL`).
> > +If one of them is marked by scheduler as unschedulable (pod condition `PodScheduled` set to `false`, the reason set to `Unschedulable`)
> > +the component will try to find a space for the addon by evicting some pods and then the scheduler will schedule the addon.
>
> s/evicting/deleting/ ? Or are we defining "eviction" at this time?
>
> Yeah, probably the introduction should say they evict and delete are
> synonymous. I don't think there's any meaningful distinction between them
> (e.g. there's no implication that something that is evicted will be
> rescheduled if it's not managed by a controller).
>
> Will the scheduler always give these pods priority? What prevents the
> deleted pods from being scheduled first?
>
> See the section below that starts "To avoid situation when Scheduler will
> schedule another pod into the space prepared for the critical addon"
>
> How does this interact with multiple schedulers? Can it delete pods from a
> non-default scheduler? Will that scheduler know not to immediately
> re-schedule in that same place?
>
> I think the mechanism described here works with multiple schedulers.
> Multiple schedulers makes it hard to know whether an evicted pod will be
> able to reschedule, but not such determination needs to be made in this
> scheme.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29195/files/a135c597401e8d9e12e19845600b8061ffdb4b7c#r71993725>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVIRPk3FOx_gOgZY0_eh2CB0vS133ks5qY7qQgaJpZM4JPqoM>
> .
>